README.md
# feathers-authentication-keystone

[![Build Status](https://travis-ci.org/virtuozzo/feathers-authentication-keystone.png?branch=master)](https://travis-ci.org/virtuozzo/feathers-authentication-keystone)
[![Code Climate](https://codeclimate.com/github/virtuozzo/feathers-authentication-keystone/badges/gpa.svg)](https://codeclimate.com/github/virtuozzo/feathers-authentication-keystone)
[![Test Coverage](https://codeclimate.com/github/virtuozzo/feathers-authentication-keystone/badges/coverage.svg)](https://codeclimate.com/github/virtuozzo/feathers-authentication-keystone/coverage)
[![Dependency Status](https://img.shields.io/david/virtuozzo/feathers-authentication-keystone.svg?style=flat-square)](https://david-dm.org/virtuozzo/feathers-authentication-keystone)
[![Download Status](https://img.shields.io/npm/dm/feathers-authentication-keystone.svg?style=flat-square)](https://www.npmjs.com/package/feathers-authentication-keystone)

> Keystone authentication strategy for feathers-authentication using Passport

## Installation

```
npm install feathers-authentication-keystone --save
```

## Documentation

Please refer to the [feathers-authentication-keystone documentation](http://docs.feathersjs.com/) for more details.


#### Default Options

```js
{
    name: 'keystone', // the name to use when invoking the authentication Strategy
    entity: 'user', // the entity that will be added to the request, socket, and hook.params. (ie. req.user, socket.user, hook.params.user)
    service: 'users', // the service to look up the entity
    passReqToCallback: true, // whether the request object should be passed to `verify`
    authUrl: 'http://<address>:<port>', // keystone endpoint with configured v3 authentication
    usernameField: 'username', // key name of username field
    passwordField: 'password', // key name of password field
    Verifier: Verifier, // A Verifier class. Defaults to the built-in one but can be a custom one. See below for details.
}
```

### Verifier

This is the verification class that receives the users identity (if verification is successful), populates the entity (normally a user) and returns both the entity and the payload. It has the following methods that can all be overridden.

```js
{
    constructor(app, options) // the class constructor
    verify(req, payload, done) // queries the configured service
}
```

#### Customizing the Verifier

The `Verifier` class can be extended so that you customize it's behavior without having to rewrite and test a totally custom local Passport implementation. Although that is always an option if you don't want use this plugin.

An example of customizing the Verifier:

```js
import keystone, { Verifier } from 'feathers-authentication-keystone';

class CustomVerifier extends Verifier {
  // The verify function has the exact same inputs and
  // return values as a vanilla passport strategy
  verify(req, payload, done) {
    // do your custom stuff. You can call internal Verifier methods
    // and reference this.app and this.options. This method must be implemented.
    done(null, payload);
  }
}

app.configure(keystone({ Verifier: CustomVerifier }));
```

## Complete Example

Here's an example of a Feathers server that uses `feathers-authentication-keystone`.

```js
const feathers = require('feathers');
const rest = require('feathers-rest');
const hooks = require('feathers-hooks');
const memory = require('feathers-memory');
const bodyParser = require('body-parser');
const errorHandler = require('feathers-errors/handler');
const auth = require('feathers-authentication');
const jwt = require('feathers-authentication-jwt');
const keystone = require('../lib/index');

// Initialize the application.
const app = feathers();

app.configure(rest())
  .configure(hooks())
  // Needed for parsing bodies (login).
  .use(bodyParser.json())
  .use(bodyParser.urlencoded({ extended: true }))
  // Configure feathers-authentication.
  .configure(auth({ secret: 'super secret' }))
  // Configure authentication by JWT token.
  .configure(jwt())
  // Keystone endpoint, this value should be taken from the config in real application.
  .configure(keystone({ authUrl: 'https://127.0.0.1:5000' }))
  .use('/users', memory())
  .use(errorHandler());

app.listen(3030);

console.log('Feathers app started on 127.0.0.1:3030');
```

## License

Copyright (c) 2017

Licensed under the [MIT license](LICENSE).


CHANGELOG.md
# Change Log

## [v0.2.0](https://github.com/virtuozzo/feathers-authentication-keystone/tree/v0.2.0) (2017-07-22)
[Full Changelog](https://github.com/virtuozzo/feathers-authentication-keystone/compare/v0.1.0...v0.2.0)

**Fixed bugs:**

- Better error reporting [\#3](https://github.com/virtuozzo/feathers-authentication-keystone/issues/3)
- Sanity errors for cases with invalid credentials and unreachable server [\#4](https://github.com/virtuozzo/feathers-authentication-keystone/pull/4) ([amisa](https://github.com/amisa))

**Merged pull requests:**

- Update plugin structure [\#5](https://github.com/virtuozzo/feathers-authentication-keystone/pull/5) ([alkurbatov](https://github.com/alkurbatov))

## [v0.1.0](https://github.com/virtuozzo/feathers-authentication-keystone/tree/v0.1.0) (2017-07-03)
**Implemented enhancements:**

- Miscellaneous fixes [\#1](https://github.com/virtuozzo/feathers-authentication-keystone/pull/1) ([alkurbatov](https://github.com/alkurbatov))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

test/index.test.js
import { expect } from 'chai';
import plugin from '../src';

describe('feathers-authentication-keystone', () => {
  it('is CommonJS compatible', () => {
    expect(typeof require('../lib')).to.equal('function');
  });

  it('basic functionality', () => {
    expect(typeof plugin).to.equal('function', 'It worked');
  });
});


src/index.js
const debug = require('debug')('feathers-authentication-keystone');
const merge = require('lodash.merge');
const omit = require('lodash.omit');
const pick = require('lodash.pick');
const DefaultVerifier = require('./verifier');
const { Strategy } = require('./passport');

const defaults = {
  name: 'keystone',
  usernameField: 'username',
  passwordField: 'password'
};

const INCLUDE_KEYS = [
  'authUrl',
  'service',
  'entity',
  'usernameField',
  'passwordField',
  'passReqToCallback'
];

const init = function init (options = {}) {
  return function keystoneAuth () {
    const app = this;
    const _super = app.setup;

    if (!app.passport) {
      throw new Error(`Can not find app.passport. Did you initialize feathers-authentication before feathers-authentication-keystone?`);
    }

    let name = options.name || defaults.name;
    let authOptions = app.get('authentication') || {};
    let keystoneOptions = authOptions[name] || {};

    let keystoneSettings = merge({}, defaults,
      pick(authOptions, INCLUDE_KEYS), keystoneOptions, omit(options, ['Verifier']));

    if (typeof keystoneSettings.authUrl === 'undefined') {
      throw new Error(`You must provide an 'authUrl' in your authentication configuration`);
    }

    let Verifier = DefaultVerifier;

    if (options.Verifier) {
      Verifier = options.Verifier;
    }

    app.setup = function () {
      let result = _super.apply(this, arguments);
      let verifier = new Verifier(app, keystoneSettings);

      if (!verifier.verify) {
        throw new Error(`Your verifier must implement a 'verify' function. It should have the same signature as a keystone passport verify callback.`);
      }

      debug('Registering keystone authentication strategy with options:', keystoneSettings);
      app.passport.use(keystoneSettings.name,
        new Strategy(keystoneSettings, verifier.verify.bind(verifier))
      );
      app.passport.options(keystoneSettings.name, keystoneSettings);

      return result;
    };
  };
};

init.Verifier = DefaultVerifier;

module.exports = init;


src/verifier.js
const debug = require('debug')('feathers-authentication-keystone:verify');

module.exports = class KeystoneVerifier {
  constructor (app, options = {}) {
    this.app = app;
    this.options = options;

    this.service = typeof options.service === 'string' ? app.service(options.service) : options.service;

    if (!this.service) {
      throw new Error(`options.service does not exist.\n\tMake sure you are passing a valid service path or service instance and it is initialized before feathers-authentication-keystone.`);
    }

    this.verify = this.verify.bind(this);
  }

  verify (req, identity, done) {
    debug('Received Keystone identity:', identity);

    let user = {
      id: identity.user.id,
      username: identity.user.name,
      token: identity.token,
      expires_at: identity.expires_at
    };

    const payload = {
      [`${this.options.entity}Id`]: identity.user.id
    };

    return done(null, user, payload);
  }
};


src/passport/strategy.js
const util = require('util');
const lookup = require('./utils').lookup;
const passport = require('passport-strategy');
const openstack = require('openstack-wrapper');

function Strategy (options, verify) {
  if (typeof options === 'function') {
    verify = options;
    options = {};
  }

  if (!options.authUrl) {
    return this.fail({ message: options.badRequestMessage || 'Missing Identity Endpoint' }, 400);
  }

  if (!verify) {
    throw new TypeError('KeystoneStrategy requires a verify callback');
  }

  this._authUrl = options.authUrl + '/v3';

  this._usernameField = options.usernameField || 'username';
  this._passwordField = options.passwordField || 'password';

  passport.Strategy.call(this);
  this.name = 'keystone';
  this._verify = verify;
  this._passReqToCallback = options.passReqToCallback;
}

util.inherits(Strategy, passport.Strategy);

// Authenticate a user using the credentials provided in the request.
Strategy.prototype.authenticate = function (req, options) {
  options = options || {};

  let username = lookup(req.body, this._usernameField) ||
    lookup(req.query, this._usernameField);
  let password = lookup(req.body, this._passwordField) ||
    lookup(req.query, this._passwordField);

  if (!username || !password) {
    return this.fail({ message: options.badRequestMessage || 'Missing credentials' }, 400);
  }

  const self = this;

  const verified = function (err, user, info) {
    if (err) {
      return self.error(err);
    }

    if (!user) {
      return self.fail(info);
    }

    return self.success(user, info);
  };

  const keystone = new openstack.Keystone(self._authUrl);

  keystone.getToken(username, password, function (err, token) {
    if (err) {
      if (err.errno === 'EHOSTUNREACH') {
        return self.fail({message: 'Unable to connect to keystone at ' + err.address + ':' + err.port}, 401);
      }

      if (err.detail && (err.detail.remoteStatusCode === 401)) {
        return self.fail({ message: 'Invalid username or password' }, 401);
      } else {
        return self.fail(err);
      }
    }

    if (self._passReqToCallback) {
      return self._verify(req, token, verified);
    }

    return self._verify(token, verified);
  });
};

module.exports = Strategy;


example/app.js
const feathers = require('feathers');
const rest = require('feathers-rest');
const hooks = require('feathers-hooks');
const memory = require('feathers-memory');
const bodyParser = require('body-parser');
const errorHandler = require('feathers-errors/handler');
const auth = require('feathers-authentication');
const jwt = require('feathers-authentication-jwt');
const keystone = require('../lib/index');

function printUserPayload () {
  return function (hook) {
    console.log('Received payload: ', hook.params.payload.user);
    return Promise.resolve(hook);
  };
}

// Initialize the application.
const app = feathers();

app.configure(rest())
  .configure(hooks())
  // Needed for parsing bodies (login).
  .use(bodyParser.json())
  .use(bodyParser.urlencoded({ extended: true }))
  // Configure feathers-authentication.
  .configure(auth({ secret: 'super secret' }))
  // Configure authentication by JWT token.
  .configure(jwt())
  // Keystone endpoint, this value should be taken from the config in real application.
  // v3 auth is used by default.
  .configure(keystone({ authUrl: 'https://127.0.0.1:5000' }))
  .use('/users', memory())
  .use(errorHandler());

// Authenticate the user using the keystone strategy and
// if successful return a token.
app.service('authentication').hooks({
  before: {
    // The credentials should be taken from the request.
    create: [auth.hooks.authenticate('keystone')]
  }
});

// Authenticate using the JWT token provided to
// the 'authentication' request.
app.service('users').hooks({
  before: {
    find: [auth.hooks.authenticate('jwt'), printUserPayload()]
  }
});

app.listen(3030);

console.log('Feathers authentication with keystone auth started on 127.0.0.1:3030');
